contractcourt/htlc_outgoing_contest_resolver.go

```diff
@@ -36,6 +35,47 @@ func newOutgoingContestResolver(res lnwallet.OutgoingHtlcResolution,
 	}
 }
 
+// Launch will call the inner resolver's launch method if the expiry height has
+// been reached, otherwise it's a no-op.
+func (h *htlcOutgoingContestResolver) Launch() error {
+	// NOTE: we don't mark this resolver as launched as the inner resolver
+	// will set it when it's launched.
+	if h.launched {
+		h.log.Tracef("already launched")
+		return nil
+	}
+
+	// If the HTLC has custom records, then for now we'll pause resolution.
+	//
+	// TODO(roasbeef): Implement resolving HTLCs with custom records
+	// (follow-up PR).
+	if len(h.htlc.CustomRecords) != 0 {
+		select { //nolint:gosimple
+		case <-h.quit:
+			return errResolverShuttingDown
+		}
+	}
+
+	h.log.Debugf("launching contest resolver...")
+
+	_, bestHeight, err := h.ChainIO.GetBestBlock()
+	if err != nil {
+		return err
+	}
+
+	if uint32(bestHeight) < h.htlcResolution.Expiry {
```

Comment:
So then the channel arb will call `Launch` for each new incoming block?

Reply:
Yeah, whenever a new block comes in, the channel arb will `Launch` all the resolvers to ensure we don't miss sending out the sweep requests. If we could instead let the channel arb manage the transformation `contest resolver -> timeout resolver`, then we don't need this, but that requires more refactor so this path is chosen.

Reply:
I wonder why we cannot do this in the resolve method ? The resolver get's notified with each block an acts accordingly.

Reply:
I believe the reasons are already clarified in the PR description,
> The original attempt is to make the resolvers subscribe to a blockbeat chan, as implemented in https://github.com/lightningnetwork/lnd/pull/8717. The second attempt is to make the resolvers also blockbeat Consumer, as implemented https://github.com/yyforyongyu/lnd/pull/8.

And more detailed considerations below.